aws/rust-runtime/aws-inlineable/src/glacier_interceptors.rs

```diff
@@ -0,0 +1,454 @@
+    fn account_id_mut(&mut self) -> &mut Option<String>;
+
+    /// Autofills the account ID with the default if not set
+    fn autofill_account_id(&mut self) {
+        let account_id = self.account_id_mut();
+        if account_id.as_deref().unwrap_or_default().is_empty() {
+            *account_id = Some(DEFAULT_ACCOUNT_ID.into());
+        }
+    }
+}
+
+/// Autofills account ID input fields with a default if no value is set
+#[derive(Debug)]
+pub(crate) struct GlacierAccountIdAutofillInterceptor<I> {
+    _phantom: PhantomData<I>,
+}
+
+impl<I> GlacierAccountIdAutofillInterceptor<I> {
+    /// Constructs a new [`GlacierAccountIdAutofillInterceptor`]
+    pub(crate) fn new() -> Self {
+        Self {
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<I: GlacierAccountId + Send + Sync + 'static> Interceptor
+    for GlacierAccountIdAutofillInterceptor<I>
+{
+    fn modify_before_serialization(
+        &self,
+        context: &mut InterceptorContext<BeforeSerialization>,
+        _cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let erased_input = context.input_mut();
+        let input: &mut I = erased_input
+            .downcast_mut()
+            .expect("typechecked at registration");
+        input.autofill_account_id();
+        Ok(())
+    }
+}
+
+/// Attaches the `x-amz-glacier-version` header to the request
+#[derive(Debug)]
+pub(crate) struct GlacierApiVersionInterceptor {
+    api_version: &'static str,
+}
+
+impl GlacierApiVersionInterceptor {
+    /// Constructs a new [`GlacierApiVersionInterceptor`] with the given API version.
+    pub(crate) fn new(api_version: &'static str) -> Self {
+        Self { api_version }
+    }
+}
+
+impl Interceptor for GlacierApiVersionInterceptor {
+    fn modify_before_signing(
+        &self,
+        context: &mut InterceptorContext<BeforeTransmit>,
+        _cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        context.request_mut().headers_mut().insert(
+            API_VERSION_HEADER,
+            HeaderValue::from_static(self.api_version),
+        );
+        Ok(())
+    }
+}
+
+/// Prevents signing from recalculating the payload's SHA-256 for operations that get the tree hashes.
+#[derive(Debug, Default)]
+pub(crate) struct GlacierSetPrecomputedSignableBodyInterceptor;
+
+impl Interceptor for GlacierSetPrecomputedSignableBodyInterceptor {
+    fn modify_before_signing(
+        &self,
+        context: &mut InterceptorContext<BeforeTransmit>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        // Get the content SHA-256 calculated during serialization (via the `RequestSerializer` Glacier customization)
+        let content_sha256 = context
+            .request()
+            .headers()
+            .get(X_AMZ_CONTENT_SHA256)
+            .ok_or_else(|| {
+                format!("{X_AMZ_CONTENT_SHA256} should have been set during serialization")
+            })?
+            .to_str()?
+            .to_owned();
```

Comment:
My gut tells me we should be handling the `or_else` after converting the possible OK value to a string to avoid allocating two strings when the header is missing. Not a big deal, though.

Reply:
There's no allocation occurring until the `to_owned` call. It's all references. The error case will allocate for the error message though, but then that's immediately bubbled up.